core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/kerberos/KeytabConfigurationHandler.java

```diff
@@ -0,0 +1,119 @@
+    private KerberosConfigService kerberosConfigService;
+
+    @Inject
+    private KerberosDetailService kerberosDetailService;
+
+    @Inject
+    private KerberosMgmtV1Endpoint kerberosMgmtV1Endpoint;
+
+    @Inject
+    private HostOrchestrator hostOrchestrator;
+
+    @Inject
+    private SecretService secretService;
+
+    @Override
+    public String selector() {
+        return EventSelectorUtil.selector(KeytabConfigurationRequest.class);
+    }
+
+    @Override
+    public void accept(Event<KeytabConfigurationRequest> keytabConfigurationRequestEvent) {
+        Long stackId = keytabConfigurationRequestEvent.getData().getResourceId();
+        Selectable response;
+        try {
+            Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+            Optional<KerberosConfig> kerberosConfigOptional = kerberosConfigService.get(stack.getEnvironmentCrn());
+            // TODO remove YARN check when FreeIPA registration is ready
+            if (!CloudPlatform.YARN.name().equals(stack.cloudPlatform())
+                    && kerberosConfigOptional.isPresent() && kerberosDetailService.isIpaJoinable(kerberosConfigOptional.get())) {
+                GatewayConfig primaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+                ServiceKeytabResponse serviceKeytabResponse = getServiceKeytabResponse(stack, primaryGatewayConfig);
+                KeytabModel keytabModel = buildKeytabModel(serviceKeytabResponse);
+                hostOrchestrator.uploadKeytabs(List.of(primaryGatewayConfig), Set.of(keytabModel),
+                        ClusterDeletionBasedExitCriteriaModel.clusterDeletionBasedModel(stackId, stack.getCluster().getId()));
+            }
+            response = new KeytabConfigurationSuccess(stackId);
+        } catch (Exception e) {
+            LOGGER.info("Error during keytab configuration, stackId: " + stackId, e);
+            response = new KeytabConfigurationFailed(stackId, e);
+        }
+        eventBus.notify(response.selector(), new Event<>(keytabConfigurationRequestEvent.getHeaders(), response));
+    }
+
+    private KeytabModel buildKeytabModel(ServiceKeytabResponse serviceKeytabResponse) {
+        String keytabInBase64 = secretService.getByResponse(serviceKeytabResponse.getKeytab());
+        byte[] keytab = Base64.getDecoder().decode(keytabInBase64.getBytes(StandardCharsets.UTF_8));
+        String principal = secretService.getByResponse(serviceKeytabResponse.getServicePrincial());
+        return new KeytabModel("CM", "/etc/cloudera-scm-server", "cmf.keytab", principal, keytab);
+    }
+
+    private ServiceKeytabResponse getServiceKeytabResponse(Stack stack, GatewayConfig primaryGatewayConfig)
+            throws Exception {
+        ServiceKeytabRequest request = new ServiceKeytabRequest();
+        request.setEnvironmentCrn(stack.getEnvironmentCrn());
+        request.setServerHostName(primaryGatewayConfig.getHostname());
+        request.setServiceName("CM");
```

Comment:
There could be other use cases that would want to use "KeytabConfigurationHandler". For example MLX.

MLX would want to use KeytabConfigurationHandler to create ketyab with administrative permissions. Having hardcoded strings like "CM" would need a rewrite if it has to be reused. Can you make these API's generic so that they can be re-used?

I'm giving this comment with the limited knowledge I have. Please ignore if it doesn't make sense.

Reply:
yes, Jamison is right. This logic is in CB for generating keytab for CM, and for CM I think the right choice is CM. Do you think we should expose this variable on our API? I'm not sure about that.

Reply:
Yep, this keytab should not be granted to any other applications, besides CM.

Reply:
yes, see: https://github.com/hortonworks/cloudbreak/pull/5591#discussion_r302852273